added option to also check for unknown query parameters

mortbauer opened the pull request:
Currently unknown query parameters are just ignored by fastapi, I feel that it should be at least possible to change the behaviour to return a validation error also for unknown.

It adresses #1190 

* not yet documented
* not yet any testing
* maybe different path to pass that option
* maybe rename to something else, maybe could be thought to generally
  check unknowns, however for models, pydantic has to be adapted

rogersjw commented:
Is there any progress on this? I see there are some merge conflicts and missing tests. Having this optional feature would be a nice to have for some work I am doing as well

mortbauer commented:
No progress from my side, I'm a bit short on time and before investing into testing and docs I thought to wait for a bit of feedback about the implementation. But I guess with more than 60 pending pull requests it is quite difficult to keep up.

sebg1 commented:
I am interested with the fix, when it will be available.

jeremyschiemann commented:
+1

hawkaa commented:
I'm interested in the fix too. I know it has been considered as an edge case by @tiangolo in a similar issue/PR, but I can explain my niece use case.

I'm implementing the "OGC API - Features - Part 1: Core" (http://docs.opengeospatial.org/is/17-069r3/17-069r3.html) which is a standard for fetching geospatial features over HTTP. In section 7.6 we read the following:

> The server SHALL respond with a response with the status code 400, if the request URI includes a query parameter that is not specified in the API definition.

This is not a big deal, but just wanted to let you guys know of an use case where this functionality would come in handy.

mortbauer committed:
added option to also check for unknown query parameters

* minimal documentation added
* tests added
* maybe different path to pass that option
* maybe rename to something else, maybe could be thought to generally
  check unknowns, however for models, pydantic has to be adapted

mortbauer committed:
fix typing for left_query_params

mortbauer committed:
fix formatting and typing

mortbauer commented:
so Í took some time to work on this again:
* rebased
* added tests
* added minimal docs
* added support to also check for unknown body params
could it be merged?

mortbauer committed:
improve test_check_unknown to hit 100% coverage

mortbauer committed:
fix style again

hawkaa commented on tests/test_check_unknown.py:
```diff
@@ -0,0 +1,66 @@
+app_check = FastAPI(check_unknown=True)
+
+client_no_check = TestClient(app_no_check)
+client_ceck = TestClient(app_check)
```
Not a big deal, but should probably be `client_check` and not `client_ceck` 👍

mortbauer replied:
yes you are right, changed that :)

mortbauer committed:
fix type in test_check_unknown

philipbjorge commented:
@mortbauer @boyaps @diogoduartec --
We'd love to have this feature for our API product.
Do you know what it will take to move this across the finish line?

Thanks -- Appreciate all the hard work you're putting into FastAPI!

gergelyhunyady commented:
@mortbauer @boyaps @diogoduartec 
I also believe this is a reasonable optional feature to add. Could serve the good in many use-cases.

As far as I understand the proposed implementation would allow users to specify `check_unknown` either for the whole service (via passing it to the `FastAPI` constructor) or alternatively they could specify it for specific `APIRouter`s only. I think this is pretty nice. Next to that I think it would be also reasonable if one could go even more fine grained by specifying the `check_unknown` behavior for specific path operations exclusively if needed (`@router.get("/users", ..., check_unknown=True)`). I didn't properly investigate if it fits well to the implementation of fastapi or if it would be hard to implement, so this is just my view from the user perspective.

Another point I would stress is the name of the parameter. I feel the `check_unknown` name is a bit too general. If the behavior strictly means only that unknown parameters will be caught and and an error response will be returned to the caller (which I think is the reasonable behavior), then I think something like `check_unknown_parameters` or maybe even better `forbid_unknown_parameters` (which would be in line with the `forbid` setting of the `extra` field in the pydantic model config).

Thanks for your work! What's your opinion on my two cents?

hawkaa commented:
@gergelyhunyady I agree on the naming part! And if we can align the naming with `pydantic` that is even better.

As for the specific path keyword, I have no strong opinions.

gguta commented:
Are there any progress on this?

cikay commented on fastapi/dependencies/utils.py:
```diff
@@ -652,7 +672,12 @@ async def request_body_to_args(
+        if received_body is not None:
+            left_body_params = set(received_body)
+        else:
+            left_body_params = set()
```
```py
left_body_params = set(received_body or {})
```

cikay commented on fastapi/dependencies/utils.py:
```diff
@@ -641,6 +660,7 @@ def request_params_to_args(
 async def request_body_to_args(
     required_params: List[ModelField],
     received_body: Optional[Union[Dict[str, Any], FormData]],
+    check_unknown: bool = False,
```
Better to name `is_unknown`.

cikay commented on fastapi/dependencies/utils.py:
```diff
@@ -571,12 +580,22 @@ async def solve_dependencies(
+    if check_unknown:
+        dependant_names = {param.name for param in dependant.query_params}
+        left_query_params.difference_update(dependant_names)
+        if outer_level:
```
Merge this condition to the upper one. Avoid nested conditions if you do not have to.

pacoispaco commented:
I am also interested in this feature. However shouldn't the default status code returned be a 404 and not a 400. Query parameters are part of the URL that identify a resource. And in the case of an unexpected query parameter the URL, by definition, would be identifying a resource that does not exist.

danhje commented:
Just noticed the application I'm currently working on has been querying an internal API with an old query parameter name for months, without anyone noticing, resulting in reduced quality of it's output. Will definitely make use of this feature when it is available!

stat-johan commented:
I would also greatly appreciate this feature

grzracz commented:
Would also deeply appreciate this.
We have a freely available API and because of "bad" query parameters, bad actors are able to bypass cache.
Instead of 
```
/assets/X/price?currency=USD
```
they are calling
```
/assets/X/price?currencyVNO1O1X114
```
adding a random string. I'd love to return a 400 when something like this is attempted.

Toby14 commented:
Would also appreciate this feature as users would know what was wrong if they make a typo or unknown param.

xsway commented:
I also definitely see a use for this. Just wanted to say that a fail on validation is not maybe needed, if there is be a way to show a warning, that would be sufficient imo.

RonC-LL commented:
I'm just here to lend my support for this feature.  Misspelling a param bit me in the butt as well. Fortunately, it's not production code or anything.

YuriiMotov commented:
@mortbauer, thanks for your contribution!

Now we can implement this logic using Pydantic models (see code example in details).
Seems that it should be enough to cover such use-cases. What do you think?

<details>

```python
from typing import Annotated
from fastapi import FastAPI, Query
from fastapi.testclient import TestClient
from pydantic import BaseModel, ConfigDict

app_check = FastAPI(check_unknown=True)

client_check = TestClient(app_check)


class BaseExtraForbid(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Item(BaseExtraForbid):
    data: str


class BodyParams(BaseExtraForbid):
    item: Item
    item2: Item


class QueryParams(BaseExtraForbid):
    param1: str
    param2: int = 0


@app_check.post("/with-extra-forbid")
async def endpoint_check(
    body: BodyParams,
    query: Annotated[QueryParams, Query()],
):
    return {
        "item": body.item,
        "item2": body.item2,
        "param1": query.param1,
        "param2": query.param2,
    }


def test_unknown():
    response = client_check.post(
        "/with-extra-forbid",
        json={"item": {"data": "myitem"}, "item2": {"data": "item2"}, "data": "any"},
        params={"param1": "qwerty", "limit": 12},
    )
    assert response.status_code == 422, response.text
    assert response.json() == {
        "detail": [
            {
                "loc": ["query", "limit"],
                "msg": "Extra inputs are not permitted",
                "type": "extra_forbidden",
                "input": "12",
            },
            {
                "loc": ["body", "data"],
                "msg": "Extra inputs are not permitted",
                "type": "extra_forbidden",
                "input": "any",
            },
        ]
    }
```

</details>
